packages/dashboard/src/contexts/DashContext/DashProvider.tsx

```diff
@@ -23,46 +37,200 @@ function DashProvider({ children }: DashProviderProps): JSX.Element {
+          // @ts-ignore
+          provider: window.ethereum
+        });
+        dispatch({
+          type: "set-decoder",
+          data: {
+            decoder,
+            decoderCompilations,
+            decoderCompilationHashes
+          }
         });
+      }
+    },
+    [dbHelper]
+  );
+
+  useEffect(() => {
+    // This obviates the need for a cleanup callback
+    if (initCalled.current) return;
+    initCalled.current = true;
+
+    const initBusClient = async () => {
+      const { busClient } = state;
+      await busClient.ready();
+      const { host, port } = busClient.options;
+      console.debug(`Connected to message bus at ws://${host}:${port}`);
+
+      // Message bus client subscribes to and handles messages
+      const subscription = busClient.subscribe({});
+      const messageHandler = async (
+        lifecycle: ReceivedMessageLifecycle<Message>
+      ) => {
+        const { message } = lifecycle;
+        if (isCliEventMessage(message)) {
+          console.debug("Received cli-event message", message);
+          if (message.payload.label === "workflow-compile-result") {
+            // Handle compilations separately because:
+            // a) Db operations are async
+            // b) Avoid duplicate work (e.g. loop, hash)
+            handleWorkflowCompileResult(
+              (message as CliEventMessage<WorkflowCompileResult>).payload.data
+            );
+          }
+        } else if (isLogMessage(message)) {
+          // Log messages do not alter state, not sending to reducer
+          console.debug(`Received log message`, message);
+          lifecycle.respond({ payload: undefined });
+        } else if (isDebugMessage(message)) {
+          // Debug messages do not alter state, not sending to reducer
+          console.debug("Received debug message", message);
+          lifecycle.respond({ payload: undefined });
+        } else {
+          // Other messages do alter state
+          dispatch({
+            type: "handle-message",
+            data: lifecycle
+          });
+        }
+      };
       subscription.on("message", messageHandler);
-    }
+    };
 
-    init();
-  }, [state]);
+    const initDecoder = async () => {
+      const { dbPromise } = state;
+      const compilationStore = await (await dbPromise).getAll("Compilation");
+      console.debug(`Compilation store`, compilationStore);
 
-  useEffect(() => {
-    dispatch({
-      type: "set-notice",
-      data: { show: !isConnected, type: "CONNECT" }
-    });
-  }, [isConnected]);
+      const decoderCompilations: State["decoderCompilations"] = new Array(
+        compilationStore.length
+      );
+      const decoderCompilationHashes: State["decoderCompilationHashes"] =
+        new Set();
+      compilationStore.forEach((row, index) => {
+        decoderCompilations[index] = row.data;
+        decoderCompilationHashes.add(row.dataHash);
+      });
+
+      const decoder = await forProject({
```

Comment:
Apologies if I missed it, but is this now using the new interface to add compilations to the decoder?

(If it's not already using this interface, it can probably be a separate task if you want, but should at least open an issue to do that switch.)

Reply:
Yea I'll leave that as a follow-up, we still re-init the decoder on new compilations.
I'll open an issue once this is merged, thanks for the reminder!